### PR TITLE
[javac] Diagnostic ranges at the class level are too aggressive

### LINKED ISSUE #299: [javac] Diagnostic ranges at the class level are too aggressive
See https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/290#issuecomment-2077135646

Diagnostic ranges at the class level are too aggressive, basically painting the whole file in yellow (eg. https://github.com/spring-projects/spring-petclinic/blob/main/src/main/java/org/springframework/samples/petclinic/owner/Owner.java):

Screenshot 2024-04-25 at 15 03 11

### PR DESCRIPTION
**Pull request by snjeza**
Fixes #299

### CODE REVIEW

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacProblemConverter.java [hunk 1]
```diff
@@ -13,36 +13,94 @@
...
+		if (diagnostic instanceof JCDiagnostic jcDiagnostic
+				&& jcDiagnostic.getDiagnosticPosition() instanceof JCClassDecl jcClassDecl) {
+			startPosition = (int) diagnostic.getPosition();
+			if (startPosition != Position.NOPOS) {
+				DiagnosticSource source = jcDiagnostic.getDiagnosticSource();
+				JavaFileObject fileObject = source.getFile();
+				try {
+					CharSequence charContent = fileObject.getCharContent(true);
+					String content = charContent.toString();
+					String name = jcClassDecl.getSimpleName().toString();
```
Comment (datho7561) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacProblemConverter.java:77]: Is the range of the name available as a child node of the jcClassDecl? I think it should be: https://github.com/openjdk/jdk/blob/33e81229bd1b4b28cf2e35f0f8f0a42a04d59c3d/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/JCTree.java#L778

**Comment by fbricon** [comment]
something's wrong with the PR, there are waayyyy too many commits

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacProblemConverter.java [hunk 2]
```diff
@@ -13,36 +13,94 @@
...
+	private static int toSeverity(Diagnostic<? extends JavaFileObject> diagnostic) {
+		return switch (diagnostic.getKind()) {
+			case ERROR -> ProblemSeverities.Error;
+			case WARNING, MANDATORY_WARNING -> ProblemSeverities.Warning;
+			case NOTE -> ProblemSeverities.Info;
+			default -> ProblemSeverities.Error;
+		};
+	}
+
+	private static JavacProblem missingSerialVersion(Diagnostic<? extends JavaFileObject> diagnostic, int problemId) {
```
Comment (fbricon) [inline org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacProblemConverter.java]: there's nothing specific to missing Serial Version UID in this method. It needs a more generic name (like classLevelProblem)

**Comment by mickaelistria** [comment]
> something's wrong with the PR, there are waayyyy too many commits

The branch was rebased; only look at the last commit, we can cherry-pick it.

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacProblemConverter.java (hunk 2 above)
Comment (snjeza) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacProblemConverter.java]: Fixed.

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacProblemConverter.java (hunk 1 above)
Comment (snjeza) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacProblemConverter.java:77]: I have used it - https://github.com/openjdk/jdk/blob/33e81229bd1b4b28cf2e35f0f8f0a42a04d59c3d/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/JCTree.java#L829

**Comment by snjeza** [comment]
> The branch was rebased; only look at the last commit, we can cherry-pick it.

I have updated the PR.

File: org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacProblemConverter.java (hunk 1 above)
Comment (datho7561) [reply org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacProblemConverter.java:77]: My bad, it seems like the Name is not an ast node, your implementation is good

**Review by datho7561 — APPROVED** [review]
Looks good and works well.

I filed two new issues based on similar error range issues.

We might need to base the error range on the type of the `getDiagnosticPosition()` node, because I'd guess that we want to highlight the class name for most of the cases where a class is specified as the diagnostic position.

**Pull request merged**